File: scripts/framework_check_v2.py
#!/usr/bin/env python3
"""Content Framework Enforcer v2 - Improved TF-IDF with multi-phrase matching.
Run: python3 scripts/framework_check.py 2>&1
"""
import re, sys

with open('src/app/blog/data.js', 'r') as f:
    content = f.read()

parts = content.split("slug: ")[1:]
posts = []
for part in parts:
    post = {}
    m = re.match(r'["\']([^"\']+)["\']', part)
    if not m: continue
    post['slug'] = m.group(1)
    def extract(p, name):
        m2 = re.search(p, part, re.DOTALL)
        if m2: post[name] = m2.group(1).strip()
    extract(r'title:\s*["\']([^"\']+)["\']', 'title')
    extract(r'date:\s*["\']([^"\']+)["\']', 'date')
    extract(r'author:\s*["\']([^"\']+)["\']', 'author')
    extract(r'metaTitle:\s*["\']([^"\']+)["\']', 'metaTitle')
    extract(r'metaDescription:\s*["\']([^"\']+)["\']', 'metaDescription')
    tm = re.search(r'tags:\s*\[(.*?)\]', part, re.DOTALL)
    if tm: post['tags'] = re.findall(r'["\']([^"\']+)["\']', tm.group(1))
    cm = re.search(r'content:\s*`([\s\S]*?)`\s*(?:,|\);|\n\s*[},])', part)
    if cm: post['content'] = cm.group(1)
    else:
        cm2 = re.search(r'content:\s*`([\s\S]*)`', part)
        if cm2: post['content'] = cm2.group(1)
        else: post['content'] = ''
    posts.append(post)

print(f"Total posts parsed: {len(posts)}", file=sys.stderr)

posts_to_check = [
    "ai-seo-2026-dhaka-experts-optimize-google-ai-chatgpt",
    "hiring-seo-expert-dhaka-better-roi-than-paid-ads",
    "how-to-choose-best-seo-expert-dhaka-15-things",
    "seo-case-study-dhaka-businesses-increased-organic-traffic",
    "seo-expert-vs-seo-agency-dhaka-which-is-right",
    "top-10-seo-mistakes-dhaka-businesses-fix",
    "watchzonebd-seo-case-study",
    "what-does-seo-expert-do-guide-business-owners",
    "complete-seo-guide-bangladesh-businesses-2026",
    "content-marketing-strategy-bangladeshi-brands-seo",
    "das-taxis-scotland-seo-case-study",
    "geo-optimization-prepare-business-ai-search",
    "google-business-profile-optimization-guide-bangladesh",
    "how-to-choose-right-seo-agency-bangladesh",
    "international-seo-bangladesh-exporters-global-buyers",
    "link-building-strategies-bangladesh-market",
    "local-seo-tips-dhaka-businesses-google-maps",
    "locksmith-dundee-seo-case-study",
    "mobile-seo-optimization-bangladesh-mobile-first-era",
    "morethanpanel-seo-case-study",
    "seo-garments-textile-industry-b2b-lead-generation",
    "seo-real-estate-developers-dhaka",
    "seo-vs-google-ads-whats-best-bangladesh-businesses",
    "smmgen-seo-case-study",
    "smmsun-seo-case-study",
    "stealth-windshield-repairs-seo-case-study",
    "technical-seo-checklist-bangladeshi-websites",
    "why-ecommerce-store-needs-seo-bangladesh",
    "why-md-kanok-miah-is-the-best-seo-expert-in-dhaka-bangladesh",
]

STOP_WORDS = {'a','an','the','how','what','why','when','where','is','are','do','does',
              'to','for','in','of','on','at','by','with','from','and','or','but','not',
              'your','its','our','their','this','that','can','will','has','have','had',
              'be','been','being','was','were','it','as','no','yes','so','if','then',
              'which','who','whom','would','could','should','may','might','up','out',
              'about','into','over','after','before','between','through','during'}

def get_keyword_candidates(title):
    """Extract multiple keyword phrase candidates from title."""
    clean = re.sub(r'[^\w\s]', ' ', title)
    words = clean.split()
    
    # Skip leading stop words
    start = 0
    for i, w in enumerate(words):
        if w.lower() not in STOP_WORDS:
            start = i
            break
    
    candidates = []
    # Try different lengths of phrases
    for length in [4, 3, 2]:
        if start + length <= len(words):
            phrase = ' '.join(words[start:start+length])
            candidates.append(phrase.lower())
    
    # Also try the first content word + SEO-specific combinations
    if start < len(words):
        first_word = words[start].lower()
        for w in words[start+1:start+4]:
            candidates.append(f"{first_word} {w.lower()}")
    
    return list(set(candidates))

def check_tfidf(content, title):
    candidates = get_keyword_candidates(title)
    content_lower = content.lower()
    
    best_count = 0
    best_keyword = ""
    for kw in candidates:
        count = len(re.findall(re.escape(kw), content_lower))
        if count > best_count:
            best_count = count
            best_keyword = kw
    
    if best_count == 0 and candidates:
        best_keyword = candidates[0]
    
    status = "✅" if best_count >= 5 else "❌"
    return status, best_keyword, f"{best_count} occurrences"

def check_entities(content, slug, title, tags):
    content_lower = content.lower()
    missing = []
    if not re.search(r'dhaka', content_lower) and not any('international' in s or 'scotland' in s or 'dundee' in s or 'uk' in s or 'export' in s for s in [slug, title.lower()]):
        missing.append('Dhaka (location)')
    # Always require Bangladesh or Bangladesh context
    if not re.search(r'bangladesh|bangladeshi', content_lower):
        missing.append('Bangladesh (location)')
    if not re.search(r'kanok miah', content_lower):
        missing.append('Kanok Miah (author/brand)')
    status = "✅" if not missing else "❌"
    if not missing:
        details = "All required entities present"
    else:
        details = f"Missing: {', '.join(missing)}"
    return status, details

def check_pillar_cluster(tags, content, slug):
    pillar_slugs = [
        'complete-seo-guide-bangladesh-businesses-2026',
        'local-seo-tips-dhaka-businesses-google-maps',
        'technical-seo-checklist-bangladeshi-websites',
    ]
    found = [s for s in pillar_slugs if f'/blog/{s}' in content]
    has_services = bool(re.search(r'/services/', content))
    has_locations = bool(re.search(r'/locations/', content))
    
    if found:
        return "✅", f"Links to: {', '.join(found)}"
    elif has_services or has_locations:
        return "⚠️", f"No pillar page link (has {'services' if has_services else ''}{'/' if has_services and has_locations else ''}{'locations' if has_locations else ''})"
    else:
        return "❌", "No pillar page link found"

def check_aeo_geo(content):
    q_words = ['How','What','Why','When','Where','Can','Do','Is','Are','Does','Which','Who','Should','Will']
    headings = re.findall(r'^#{2,3}\s+.*$', content, re.MULTILINE)
    q_heads = []
    for h in headings:
        h_text = h.lstrip('#').strip()
        fw = h_text.split()[0] if h_text.split() else ''
        if fw in q_words or fw.rstrip('?:') in q_words:
            q_heads.append(h_text)
    count = len(q_heads)
    status = "✅" if count >= 2 else "❌"
    if q_heads:
        labels = q_heads[:3]
        if len(q_heads) > 3: labels.append(f"+{len(q_heads)-3} more")
        details = f"{count}: {'; '.join(labels)}"
    else:
        details = "0 question headings"
    return status, details

def check_internal_linking(content, slug):
    blog = [l for l in re.findall(r'/blog/([^"\')\]\s]+)', content) if l != slug]
    svc = re.findall(r'/services/([^"\')\]\s]+)', content)
    loc = re.findall(r'/locations/([^"\')\]\s]+)', content)
    ind = re.findall(r'/industries/([^"\')\]\s]+)', content)
    total = len(blog) + len(svc) + len(loc) + len(ind)
    status = "✅" if total >= 3 else "❌"
    types = []
    if blog: types.append(f"{len(blog)} posts")
    if svc: types.append(f"{len(svc)} services")
    if loc: types.append(f"{len(loc)} locations")
    if ind: types.append(f"{len(ind)} industries")
    return status, f"{total} total ({', '.join(types)})"

def check_schema(post):
    missing = []
    for field in ['title','excerpt','date','metaTitle','metaDescription']:
        if not post.get(field): missing.append(field)
    status = "✅" if not missing else "❌"
    if not missing:
        details = "All fields set"
    else:
        details = f"Missing: {', '.join(missing)}"
    return status, details

# Run
print("=" * 80)
print("CONTENT FRAMEWORK ENFORCEMENT REPORT — kanokmiah.com.bd")
print(f"Generated: July 15, 2026")
print("=" * 80)

all_passed = True
summary_rows = []

for target_slug in posts_to_check:
    post = next((p for p in posts if p['slug'] == target_slug), None)
    if not post:
        print(f"\n{'='*60}\n## Post: {target_slug}\n⚠️ NOT FOUND\n")
        continue
    
    title = post.get('title', '')
    content = post.get('content', '')
    tags = post.get('tags', [])
    
    print(f"\n{'='*60}")
    print(f"## Post: {target_slug}")
    print(f"Title: {title}")
    print(f"{'='*60}")
    print("| Check | Status | Details |")
    print("|-------|--------|---------|")
    
    results = {}
    
    st, kw, sd = check_tfidf(content, title)
    results['TF-IDF'] = (st, f"Keyword: '{kw}' — {sd}")
    print(f"| TF-IDF | {st} | {results['TF-IDF'][1]} |")
    
    st, sd = check_entities(content, target_slug, title, tags)
    results['Entities'] = (st, sd)
    print(f"| Entities | {st} | {sd} |")
    
    st, sd = check_pillar_cluster(tags, content, target_slug)
    results['Pillar Link'] = (st, sd)
    print(f"| Pillar | {st} | {sd} |")
    
    st, sd = check_aeo_geo(content)
    results['AEO/GEO'] = (st, sd)
    print(f"| AEO/GEO | {st} | {sd} |")
    
    st, sd = check_internal_linking(content, target_slug)
    results['Internal Links'] = (st, sd)
    print(f"| Internal Links | {st} | {sd} |")
    
    st, sd = check_schema(post)
    results['Schema'] = (st, sd)
    print(f"| Schema | {st} | {sd} |")
    
    # Fix instructions
    fixes = []
    if results['TF-IDF'][0] == '❌':
        fixes.append(f"- Boost primary keyword '{kw}' to >=5 occurrences")
    if results['Entities'][0] == '❌':
        fixes.append(f"- Add missing: {results['Entities'][1].replace('Missing: ', '')}")
    if results['Pillar Link'][0] == '❌':
        fixes.append("- Add link to pillar page (/blog/complete-seo-guide-bangladesh-businesses-2026)")
    if results['AEO/GEO'][0] == '❌':
        fixes.append("- Add >=2 question-based headings (How, What, Why...)")
    if results['Internal Links'][0] == '❌':
        fixes.append("- Add >=3 internal links (blog posts, services, locations, industries)")
    if results['Schema'][0] == '❌':
        fixes.append(f"- Set schema fields: {results['Schema'][1].replace('Missing: ', '')}")
    
    if fixes:
        print("\n### Fix instructions:")
        for f in fixes: print(f)
        all_passed = False
    else:
        print("\n### ✅ All checks passed!")
    
    # Store for summary
    checks_pass = all(r[0] == '✅' for r in results.values())
    summary_rows.append((target_slug, '✅ ALL PASS' if checks_pass else '❌ NEEDS FIXES'))
    print()

print("=" * 80)
print("SUMMARY")
print("=" * 80)
for slug, status in summary_rows:
    print(f"  {status} — {slug}")
print("=" * 80)
if all_passed:
    print("OVERALL: ✅ All 29 posts pass all framework checks")
else:
    # Count how many pass/fail
    pass_count = sum(1 for _, s in summary_rows if 'PASS' in s)
    fail_count = 29 - pass_count
    print(f"OVERALL: ❌ {fail_count} posts need fixes | {pass_count} posts pass all checks")
print("=" * 80)

File: scripts/framework-check-v2.mjs
#!/usr/bin/env node
/**
 * Content Framework Enforcer v2 — handles Bengali/English bilingual content.
 */
import { readFileSync, writeFileSync } from 'fs';

const DATA_JS = new URL('../src/app/blog/data.js', import.meta.url).pathname;
const SLUGS_FILE = '/tmp/modified_slugs.txt';

const PILLAR_SLUGS = new Set([
  'complete-seo-guide-bangladesh-businesses-2026',
  'local-seo-dhaka-google-maps-ranking',
  'seo-trends-2026-ai-geo-future',
  'technical-seo-core-web-vitals-optimization',
  'seo-bangla-beginners-guide-google-ranking',
  'seo-vs-google-ads-bangladesh-business',
  'seo-for-law-firms-bangladesh',
  'seo-for-startups-bangladesh',
  'seo-for-fitness-gyms-bangladesh',
]);

const PILLAR_SERVICES = [
  '/services/', '/services/seo', '/services/local-seo',
  '/services/ecommerce-seo', '/services/technical-seo',
  '/services/content-marketing',
];

/** Extract primary keyword phrase from title (Bangla + English aware) */
function extractPrimaryKeyword(title) {
  if (!title) return '';

  // Split on colon — take the part BEFORE colon as the primary keyword zone
  const colonIdx = title.indexOf(':');
  const beforeColon = colonIdx > -1 ? title.slice(0, colonIdx).trim() : title;

  // Strip leading fluff
  const cleaned = beforeColon.replace(/^(Complete|Ultimate|Expert|Best|Top|Affordable|Professional|Comprehensive|সহজ|শ্রেষ্ঠ|সেরা|উত্তম|পূর্ণ|সম্পূর্ণ)\s+/i, '');

  // For English titles, grab the first meaningful 2-3 word phrase
  const engStop = / (Guide|Tips|Strategies|Checklist|Techniques|Optimization|Benefits|Mistakes|Services|Agency|Company|Expert|Specialist|Consultant|Tools|Ideas|Examples|Case Study|Checklist|মethods|Process|Steps|Framework|Roadmap|In |For |to |for |in |of |the |and |with |your |a |an |& |-|–)/i;
  const engMatch = cleaned.match(engStop);
  if (engMatch && engMatch.index > 0) {
    return cleaned.slice(0, engMatch.index).trim();
  }

  // For English, return first 3 words
  const engWords = cleaned.split(' ').filter(w => /[a-zA-Z]/.test(w));
  if (engWords.length >= 2) return engWords.slice(0, Math.min(3, engWords.length)).join(' ');

  // For Bangla/Bengali, return everything before colon (or first 5 words)
  const bengaliParts = cleaned.split(/[\s,]+/).filter(Boolean);
  return bengaliParts.slice(0, Math.min(5, bengaliParts.length)).join(' ');
}

/** Count occurrences of a keyword in text, with Bangla-aware fallback */
function countOccurrences(text, keyword) {
  if (!keyword || keyword.length < 2) return 0;
  const escaped = keyword.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const matches = text.match(new RegExp(escaped, 'gi'));
  return matches ? matches.length : 0;
}

/** Count question-based headings — supports Bangla + English */
function countQuestionHeadings(content) {
  // English question words
  const enWords = ['How', 'What', 'Why', 'When', 'Where', 'Can', 'Do', 'Is', 'Are', 'Does', 'Which', 'Who', 'Will', 'Has', 'Have', 'Should', 'Could', 'Would', 'May', 'Might'];
  // Bengali question words
  const bnWords = ['কী', 'কেন', 'কখন', 'কোথায়', 'কীভাবে', 'কিভাবে', 'কোন', 'কোনটি', 'কি', 'হয়ে', 'হবে', 'হয়', 'কতো', 'কত', 'কার', 'কারা'];

  const allWords = [...enWords, ...bnWords];
  const pattern = new RegExp(`^#{2,3}\\s+(${allWords.join('|')})\\b`, 'gim');
  const matches = content.match(pattern);
  return matches ? matches.length : 0;
}

/** Count internal links (same-domain markdown links) */
function countInternalLinks(content) {
  const linkPattern = /\[([^\]]*)\]\((\/[^)]*)\)/g;
  const matches = content.matchAll(linkPattern);
  let count = 0;
  for (const m of matches) {
    const url = m[2];
    if (url.startsWith('//') || url.startsWith('http') || url === '#' || url.startsWith('#') || url.startsWith('mailto:')) continue;
    count++;
  }
  return count;
}

/** Check pillar link */
function checkPillarLink(content, slug) {
  if (PILLAR_SLUGS.has(slug)) return { pass: true, detail: `Self: pillar page (${slug})` };
  for (const pslug of PILLAR_SLUGS) {
    if (content.includes(`/blog/${pslug}`)) return { pass: true, detail: `Links to pillar: ${pslug}` };
  }
  for (const svc of PILLAR_SERVICES) {
    if (content.includes(svc)) return { pass: true, detail: `Links to service: ${svc}*` };
  }
  return { pass: false, detail: 'No pillar page or service link found' };
}

/** Determine pillar topic from tags */
function determinePillarTopic(tags) {
  const tagStr = tags.join(' ').toLowerCase();
  if (tagStr.includes('local') || tagStr.includes('maps') || tagStr.includes('gbp') || tagStr.includes('google business')) return 'Local SEO';
  if (tagStr.includes('technical') || tagStr.includes('core web') || tagStr.includes('crawl') || tagStr.includes('schema') || tagStr.includes('structured data') || tagStr.includes('canonical') || tagStr.includes('robots') || tagStr.includes('sitemap') || tagStr.includes('redirect') || tagStr.includes('https') || tagStr.includes('hreflang')) return 'Technical SEO';
  if (tagStr.includes('trend') || tagStr.includes('2026') || tagStr.includes('future') || tagStr.includes('ai') || tagStr.includes('geo')) return 'SEO Trends & GEO';
  if (tagStr.includes('ecommerce') || tagStr.includes('shopify') || tagStr.includes('daraz')) return 'E-commerce SEO';
  if (tagStr.includes('link building') || tagStr.includes('backlink')) return 'Link Building';
  if (tagStr.includes('keyword') || tagStr.includes('search intent')) return 'Keyword Research';
  if (tagStr.includes('content') || tagStr.includes('blog') || tagStr.includes('writing')) return 'Content Marketing';
  if (tagStr.includes('for ') || tagStr.includes('industry') || tagStr.includes('garment') || tagStr.includes('law') || tagStr.includes('fitness') || tagStr.includes('startup') || tagStr.includes('ngo') || tagStr.includes('real estate') || tagStr.includes('hotel') || tagStr.includes('restaurant') || tagStr.includes('education') || tagStr.includes('healthcare') || tagStr.includes('youtube') || tagStr.includes('podcast') || tagStr.includes('mobile app')) return 'Industry SEO';
  if (tagStr.includes('schema') || tagStr.includes('faq') || tagStr.includes('howto') || tagStr.includes('breadcrumb') || tagStr.includes('json-ld') || tagStr.includes('rich snippet')) return 'Structured Data';
  if (tagStr.includes('international') || tagStr.includes('hreflang') || tagStr.includes('global')) return 'International SEO';
  return 'General SEO';
}

/** Entity coverage check — Bangla + English aware */
function checkEntityCoverage(content, title, slug, tags) {
  const tagStr = tags.join(' ').toLowerCase();
  const text = content.toLowerCase();
  const missing = [];

  // Location entities (check both English and Bengali forms)
  const needsBangladesh = slug.includes('bangladesh') || slug.includes('bd') || tagStr.includes('bangladesh');
  if (needsBangladesh) {
    if (!text.includes('bangladesh') && !text.includes('বাংলাদেশ') && !text.includes('বাংলাদেশী') && !text.includes('বাংলাদেশি')) {
      missing.push('Bangladesh/বাংলাদেশ');
    }
  }
  const needsDhaka = slug.includes('dhaka') || tagStr.includes('dhaka') || slug.includes('dhaka');
  if (needsDhaka) {
    if (!text.includes('dhaka') && !text.includes('ঢাকা') && !text.includes('ঢাকায়')) {
      missing.push('Dhaka/ঢাকা');
    }
  }

  // Industry/service entities (bilingual)
  const entityPairs = [
    { tag: ['garment','textile','rmg'], names: ['garment','textile','rmg','গার্মেন্টস','টেক্সটাইল'], label: 'Garments/Textile' },
    { tag: ['law','legal','attorney'], names: ['law','legal','attorney','advocate','আইন','আইনি','লিগাল'], label: 'Law/Legal' },
    { tag: ['fitness','gym','workout'], names: ['fitness','gym','workout','ফিটনেস','জিম'], label: 'Fitness/Gym' },
    { tag: ['startup','start-up'], names: ['startup','start-up','স্টার্টআপ','স্টার্ট-আপ'], label: 'Startup' },
    { tag: ['ecommerce','e-commerce','shopify','daraz','online store'], names: ['ecommerce','e-commerce','online store','shop','ই-কমার্স','অনলাইন','দারাজ','shopify'], label: 'E-commerce' },
    { tag: ['hotel','resort','lodging'], names: ['hotel','resort','lodging','হোটেল','রিসোর্ট'], label: 'Hotel/Resort' },
    { tag: ['real estate','property','apartment'], names: ['real estate','property','apartment','রিয়েল এস্টেট','প্রপার্টি'], label: 'Real Estate' },
    { tag: ['ngo','non-profit','nonprofit','charity'], names: ['ngo','non-profit','nonprofit','charity','এনজিও','দাতব্য'], label: 'NGO/Non-profit' },
    { tag: ['youtube','video'], names: ['youtube','video','ইউটিউব','ভিডিও'], label: 'YouTube/Video' },
    { tag: ['podcast','audio'], names: ['podcast','audio','পডকাস্ট'], label: 'Podcast' },
    { tag: ['mobile app','mobile application'], names: ['app','mobile application','মোবাইল অ্যাপ','অ্যাপ'], label: 'Mobile App' },
    { tag: ['education','educational','student','university','college','school'], names: ['education','student','university','college','school','শিক্ষা','বিশ্ববিদ্যালয়','কলেজ','স্কুল'], label: 'Education' },
    { tag: ['healthcare','health','medical','doctor','hospital','clinic'], names: ['health','medical','doctor','hospital','clinic','স্বাস্থ্য','চিকিৎসা','হাসপাতাল','ক্লিনিক'], label: 'Healthcare' },
    { tag: ['restaurant','food','cafe','dining'], names: ['restaurant','food','cafe','dining','রেস্টুরেন্ট','খাবার','ক্যাফে'], label: 'Restaurant/Food' },
    { tag: ['seo','search engine optimization'], names: ['seo','search engine optimization','search engine optimisation','এসইও','সার্চ ইঞ্জিন'], label: 'SEO' },
    { tag: ['local seo','local search','google maps'], names: ['local seo','local search','google maps','লোকাল','স্থানীয়'], label: 'Local SEO' },
    { tag: ['content marketing','content','blog','article'], names: ['content','blog','article','কন্টেন্ট','ব্লগ','আর্টিকেল'], label: 'Content' },
    { tag: ['schema','structured data','rich snippet','faq','howto','breadcrumb','json-ld'], names: ['schema','structured data','rich snippet','স্কিমা','স্ট্রাকচারড ডাটা','faq','howto'], label: 'Schema/Structured Data' },
    { tag: ['link building','backlink','anchor text'], names: ['link building','backlink','anchor text','লিংক','ব্যাকলিংক','লিংক বিল্ডিং'], label: 'Link Building' },
    { tag: ['technical','core web','page speed','crawl','index'], names: ['technical','crawl','index','page speed','core web','টেকনিক্যাল','ক্রল','ইনডেক্স'], label: 'Technical SEO' },
  ];

  for (const pair of entityPairs) {
    const tagMatch = pair.tag.some(t => tagStr.includes(t) || slug.includes(t));
    if (!tagMatch) continue;
    const found = pair.names.some(n => text.includes(n));
    if (!found) missing.push(pair.label);
  }

  if (missing.length === 0) return { pass: true, detail: 'All expected entities found' };
  return { pass: false, detail: `Missing: ${missing.join(', ')}` };
}

// ── Main ────────────────────────────────────────────────────────────────────

async function main() {
  const posts = (await import(DATA_JS)).default;
  const modifiedSlugs = readFileSync(SLUGS_FILE, 'utf-8')
    .split('\n').map(s => s.trim()).filter(Boolean);

  console.log(`Loaded ${posts.length} total posts, ${modifiedSlugs.length} modified slugs\n`);

  const reportParts = [];
  let totalPass = 0, totalFail = 0, totalChecks = 0;
  let postsWithIssues = 0, postsClean = 0;

  for (const slug of modifiedSlugs) {
    const post = posts.find(p => p.slug === slug);
    if (!post) {
      reportParts.push(`## Post: ${slug}\n⚠️ Not found in data.js\n`);
      continue;
    }

    const { title, excerpt, date, tags = [], content = '' } = post;
    const keyword = extractPrimaryKeyword(title);
    const occurrences = countOccurrences(content, keyword);
    const questionHeadings = countQuestionHeadings(content);
    const internalLinks = countInternalLinks(content);

    // A. TF-IDF
    const tfidfPass = occurrences >= 5;
    if (tfidfPass) totalPass++; else totalFail++; totalChecks++;

    // B. Entities
    const entityCheck = checkEntityCoverage(content, title, slug, tags);
    if (entityCheck.pass) totalPass++; else totalFail++; totalChecks++;

    // C. Pillar
    const pillarCheck = checkPillarLink(content, slug);
    const pillarTopic = determinePillarTopic(tags);
    if (pillarCheck.pass) totalPass++; else totalFail++; totalChecks++;

    // D. AEO/GEO
    const aeoPass = questionHeadings >= 2;
    if (aeoPass) totalPass++; else totalFail++; totalChecks++;

    // E. Internal Links
    const linkingPass = internalLinks >= 3;
    if (linkingPass) totalPass++; else totalFail++; totalChecks++;

    // F. Schema
    const missingSchema = [];
    if (!title) missingSchema.push('title');
    if (!excerpt) missingSchema.push('excerpt');
    if (!date) missingSchema.push('date');
    const schemaPass = missingSchema.length === 0;
    if (schemaPass) totalPass++; else totalFail++; totalChecks++;

    const hasIssues = !tfidfPass || !entityCheck.pass || !pillarCheck.pass || !aeoPass || !linkingPass || !schemaPass;
    if (hasIssues) postsWithIssues++; else postsClean++;

    const fixInstructions = [];
    if (!tfidfPass) fixInstructions.push(`- **TF-IDF**: Increase \`${keyword}\` occurrences (${occurrences} → ≥5). Use the term naturally in headings and body text.`);
    if (!entityCheck.pass) fixInstructions.push(`- **Entities**: ${entityCheck.detail}. Add missing terms at least once.`);
    if (!pillarCheck.pass) fixInstructions.push(`- **Pillar Link**: Add a link to the "${pillarTopic}" pillar page (\`/blog/…\` or \`/services/…\`).`);
    if (!aeoPass) fixInstructions.push(`- **AEO/GEO**: Add more question headings (${questionHeadings} → ≥2). Use How/What/Why or Bengali question words (কী/কেন/কীভাবে).`);
    if (!linkingPass) fixInstructions.push(`- **Internal Links**: Add more (${internalLinks} → ≥3). Link to other posts, services, or location pages.`);
    if (!schemaPass) fixInstructions.push(`- **Schema**: Missing: ${missingSchema.join(', ')}.`);

    reportParts.push(`## Post: ${slug}
| Check | Status | Details |
|-------|--------|---------|
| TF-IDF: \`${keyword}\` | ${tfidfPass ? '✅' : '❌'} | ${occurrences} occurrences |
| Entities | ${entityCheck.pass ? '✅' : '❌'} | ${entityCheck.detail} |
| Pillar Link | ${pillarCheck.pass ? '✅' : '❌'} | ${pillarTopic}: ${pillarCheck.detail} |
| AEO/GEO | ${aeoPass ? '✅' : '❌'} | ${questionHeadings} question headings |
| Internal Links | ${linkingPass ? '✅' : '❌'} | ${internalLinks} total |
| Schema Ready | ${schemaPass ? '✅' : '❌'} | ${missingSchema.length ? `Missing: ${missingSchema.join(', ')}` : 'All fields set'}

### ${hasIssues ? '⚠️ Fix instructions:' : '✅ All checks passed'}
${fixInstructions.length ? fixInstructions.join('\n') : 'No fixes needed.'}`);
  }

  const pct = totalChecks ? Math.round(totalPass / totalChecks * 100) : 0;
  const summary = `# Content Framework Report — kanokmiah.com.bd
**Date**: ${new Date().toISOString().slice(0, 10)}
**Period**: Last 48 hours
**Posts checked**: ${modifiedSlugs.length}
**Posts with issues**: ${postsWithIssues}
**Posts clean**: ${postsClean}
**Checks passed**: ${totalPass}/${totalChecks} (${pct}%)
**Checks failed**: ${totalFail}
---

${reportParts.join('\n\n')}
`;

  const reportPath = '/tmp/framework-report-v2.md';
  writeFileSync(reportPath, summary, 'utf-8');
  console.log(summary);
}

main().catch(err => { console.error('Fatal:', err); process.exit(1); });
